crates/transpiler/src/passes/sabre/route.rs

```diff
@@ -11,68 +11,357 @@
+    /// device.  If the device was subset (such as for disjoint handling), use [rebuild_onto] with
+    /// suitable mappings back to the full-width [PhysicalQubit] instances instead.
+    pub fn rebuild(&self) -> PyResult<DAGCircuit> {
+        let num_swaps = self.swap_count();
+        let dag = self.dag.physical_empty_like_with_capacity(
+            self.num_qubits(),
+            self.dag.num_ops() + num_swaps,
+            self.dag.dag().edge_count() + 2 * num_swaps,
+        )?;
+        self.rebuild_onto(dag, |q| q)
+    }
+
+    /// Construct the routed [DAGCircuit] from the result, placing the operations onto an existing
+    /// [DAGCircuit]
+    ///
+    /// `dag` should be a physical circuit that is the full width of the device, and already contain
+    /// all the classical data that is referenced by this circuit (clbits, classical registers,
+    /// vars, stretches, global phase, metadata, etc).  The target [DAGCircuit] might have more
+    /// qubits than the layouts in this [RoutingResult], for example if the device was subset for
+    /// disjoint handling.  Use `map_fn` to map the "physical qubits" in this component to the
+    /// actual physical qubits they refer to.
+    pub fn rebuild_onto(
+        &self,
+        dag: DAGCircuit,
+        map_fn: impl Fn(PhysicalQubit) -> PhysicalQubit,
+    ) -> PyResult<DAGCircuit> {
+        let apply_swap = |swap: &[PhysicalQubit; 2],
+                          layout: &mut NLayout,
+                          dag: &mut DAGCircuitBuilder|
+         -> PyResult<NodeIndex> {
+            layout.swap_physical(swap[0], swap[1]);
+            let swap = PackedInstruction::from_standard_gate(
+                StandardGate::Swap,
+                None,
+                dag.insert_qargs(&[Qubit(map_fn(swap[1]).0), Qubit(map_fn(swap[0]).0)]),
+            );
+            dag.push_back(swap)
+        };
+        // The size here is pretty arbitrary, providing it can fit at least 2q operations in.
+        let mut apply_scratch = Vec::with_capacity(4);
+        let mut apply_op = |inst: &PackedInstruction,
+                            layout: &NLayout,
+                            dag: &mut DAGCircuitBuilder|
+         -> PyResult<NodeIndex> {
+            apply_scratch.clear();
+            for qubit in self.dag.get_qargs(inst.qubits) {
+                apply_scratch.push(Qubit(map_fn(VirtualQubit(qubit.0).to_phys(layout)).0));
+            }
+            let new_inst = PackedInstruction {
+                qubits: dag.insert_qargs(&apply_scratch),
+                ..inst.clone()
+            };
+            dag.push_back(new_inst)
+        };
+
+        let mut dag = dag.into_builder();
+        let mut layout = self.initial_layout.clone();
+        let mut blocks = self.control_flow.iter();
+        for node in &self.sabre.initial {
+            let NodeType::Operation(inst) = &self.dag[*node] else {
+                panic!("sabre DAG should only contain op nodes");
```

Comment:
This would drive me crazy if I ever encountered this.

```suggestion
                panic!("Sabre DAG should only contain op nodes");
```

Reply:
Done in 7d73ed6056.